Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -69,46 +170,85 @@ class CDMPrivateOpenCDM : public CDMPrivate {
 
 private:
     String m_openCDMKeySystem;
-    media::OpenCdm m_openCDM;
+    // This owns OCDM Accessor and passes a bare pointer further because it's owned by CDMInstance
+    // which lives as long as any MediaKeySession lives.
+    ScopedOCDMAccessor m_openCDMAccessor;
 };
 
-class CDMInstanceOpenCDM::Session : public ThreadSafeRefCounted<CDMInstanceOpenCDM::Session> {
+class CDMInstanceOpenCDM::Session : public ThreadSafeRefCounted<CDMInstanceOpenCDM::Session>, public OCDMSessionCallbacksNotifier::Observer {
 public:
-    static Ref<Session> create(const media::OpenCdm& source, Ref<WebCore::SharedBuffer>&& initData);
+    using ChallengeGeneratedCallback = Function<void(Session*)>;
+    using KeyUpdatedCallback = Function<void(Session*, OCDMKeyStatus, const String&)>;
+    using SessionChangedCallback = Function<void(Session*, bool success, const String&)>;
 
-    bool isValid() const { return m_isValid; }
+    static Ref<Session> create(OpenCDMAccessor&, const String&, const char*, Ref<WebCore::SharedBuffer>&&, LicenseType, Ref<WebCore::SharedBuffer>&&);
+    ~Session() override;
+
+    bool isValid() const { return m_session.get() && m_message && !m_message->isEmpty(); }
+    const String& id() const { return m_id; }
     Ref<SharedBuffer> message() const { ASSERT(m_message); return Ref<SharedBuffer>(*m_message.get()); }
     bool needsIndividualization() const { return m_needsIndividualization; }
     const Ref<WebCore::SharedBuffer>& initData() const { return m_initData; }
-    std::pair<media::OpenCdm::KeyStatus, String> update(const uint8_t* data, unsigned length);
-    std::pair<bool, String> load();
-    std::pair<bool, String> remove();
-    bool close() { return !m_session.Close(); }
-    media::OpenCdm::KeyStatus lastStatus() const { return m_lastStatus; }
+    void generateChallenge(ChallengeGeneratedCallback&&);
+    void update(const uint8_t*, unsigned, KeyUpdatedCallback&&);
+    void load(SessionChangedCallback&&);
+    void remove(SessionChangedCallback&&);
+    bool close() { return !opencdm_session_close(m_session.get()); }
+    OCDMKeyStatus lastStatus() const { return m_lastStatus; }
```

Comment:
yeah but the API requires key to get the status so in exchange of caching the status the key would have to be stored just to be used for status retrieval.

Reply:
It's a bit of a pity. What's up with key ids? Isn't there any way to extract key ids and use that to retrieve the key status? (I think the spec mandates that keys should be able to be represented by key ids. It would be nice to use this instead. If it is not possible, I guess we'll have to live with the `m_lastStatus` hack, considering that we should keep it updated (expiration, etc.) and this seems to be true because of the `keyUpdatedCallback`?

Reply:
You're right. The API needs key id but it still makes it not that useful, no?
In current implementation the status will be updated on a callback fired as a response to update() call.

Reply:
Hm. It seems it doesn't matter what I pass as keyID there (I currently pass update request reponse there which is wrong) yet the API ignore that and returns the status. OFC it'd be wrong to depend on that still I do not see a way to get hold of a key or keyid for that matter so maybe it's the API which is broken? It'd make sense if a ocdm session could have multiple keys but I'm not sure to be honest if it's possible.

Reply:
If I could depend on OCDM to provide the status (which it currently seems to cache) without the need to give anything except OCDM session I could not only get rid of the `m_lastStatus` but also change `PendingKeyUpdate` structure to a callback